src/db.c

```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
-  char *query;
+  plversion++;
+  ret = snprintf(version, sizeof(version), "%d", plversion);
+  if (ret >= sizeof(version))
+    {
+      DPRINTF(E_LOG, L_DB, "Error incrementing queue version. Could not convert version to string: %d\n", plversion);
+      db_transaction_rollback();
+      return;
+    }
 
-  query = sqlite3_mprintf(Q_TMPL, path);
+  ret = db_admin_update("plversion", version);
+  if (ret < 0)
+    {
+      DPRINTF(E_LOG, L_DB, "Error incrementing queue version. Could not update version in admin table: %d\n", plversion);
+      db_transaction_rollback();
+      return;
+    }
 
-  return db_query_run(query, 1, 0);
-#undef Q_TMPL
+  db_transaction_end();
+
+  listener_notify(LISTENER_PLAYLIST);
 }
 
-int
-db_watch_delete_bymatch(char *path)
+void
+db_queue_update_icymetadata(int id, char *artist, char *album)
 {
-#define Q_TMPL "DELETE FROM inotify WHERE path LIKE '%q/%%';"
+#define Q_TMPL "UPDATE queue SET artist = TRIM(%Q), album = TRIM(%Q) WHERE id = %d;"
   char *query;
 
-  query = sqlite3_mprintf(Q_TMPL, path);
+  if (id == 0)
+    return;
+
+  query = sqlite3_mprintf(Q_TMPL, artist, album, id);
+  if (!query)
+    {
+      DPRINTF(E_LOG, L_DB, "Out of memory for query string\n");
+
+      return;
+    }
+
+  db_query_run(query, 1, 0);
+  queue_inc_version_and_notify();
 
-  return db_query_run(query, 1, 0);
 #undef Q_TMPL
 }
 
-int
-db_watch_delete_bycookie(uint32_t cookie)
-{
-#define Q_TMPL "DELETE FROM inotify WHERE cookie = %" PRIi64 ";"
+static int
+queue_add_mediafileinfo(struct db_media_file_info *dbmfi, int pos, int shuffle_pos)
+{
+#define Q_TMPL "INSERT INTO queue "							\
+		    "(id, file_id, song_length, data_kind, media_kind, "		\
+		    "pos, shuffle_pos, path, virtual_path, title, "			\
+		    "artist, album_artist, album, genre, songalbumid, "			\
+		    "time_modified, artist_sort, album_sort, album_artist_sort, year, "	\
+		    "track, disc)" 							\
+		"VALUES"                                           			\
+		    "(NULL, %s, %s, %s, %s, "						\
+		    "%d, %d, %Q, %Q, %Q, "						\
+		    "%Q, %Q, %Q, %Q, %s, "						\
+		    "%s, %Q, %Q, %Q, %s, "						\
+		    "%s, %s);"
+
   char *query;
+  int ret;
 
-  if (cookie == 0)
-    return -1;
+  query = sqlite3_mprintf(Q_TMPL,
+			  dbmfi->id, dbmfi->song_length, dbmfi->data_kind, dbmfi->media_kind,
+			  pos, pos, dbmfi->path, dbmfi->virtual_path, dbmfi->title,
+			  dbmfi->artist, dbmfi->album_artist, dbmfi->album, dbmfi->genre, dbmfi->songalbumid,
+			  dbmfi->time_modified, dbmfi->artist_sort, dbmfi->album_sort, dbmfi->album_artist_sort, dbmfi->year,
+			  dbmfi->track, dbmfi->disc);
+  ret = db_query_run(query, 1, 0);
 
-  query = sqlite3_mprintf(Q_TMPL, (int64_t)cookie);
+  return ret;
 
-  return db_query_run(query, 1, 0);
 #undef Q_TMPL
 }
 
-static int
-db_watch_get_byquery(struct watch_info *wi, char *query)
+/*
+ * Adds the files matching the given query to the queue after the item with the given item id
```

Comment:
Alright. I'm not an Eclipse user myself, but isn't this behavior odd? I mean most of the header files in /usr/include include documentation, so does Eclipse not show that? Like for the library headers, a caller shouldn't need to look in the implementation for anything.